// be-utils/lib/utils/HTTPRequest.js
/**
 *  Creator: Christian Hotz
 *  Company: hydra newmedia GmbH
 *  Date: 06.06.16
 *
 *  Copyright hydra newmedia GmbH
 */
"use strict";
/**
 *  Imports
 */
var request = require('request');
var Error_1 = require('./error/Error');
/**
 *
 * @param method {string}
 * @param uri {string}
 * @param headers {any}
 * @param payload {object|function}
 * @param callback {function}
 */
var HTTPRequest = (function () {
    function HTTPRequest(method, uri, headers, payload) {
        var options = {
            uri: uri,
            method: method
        };
        if (headers) {
            options.headers = headers;
        }
        if (payload) {
            options.json = payload;
        }
        this.options = options;
    }
    HTTPRequest.prototype.send = function (callback) {
        callback = (typeof callback === 'function' ? callback : null);
        request(this.options, function (err, res) {
            if (err) {
                return callback(new Error_1.Error(err.code || -1, 'httpRequest error', 'Request failed', 'httpRequest.sendRequest', err));
            }
            callback(null, res);
        });
    };
    return HTTPRequest;
}());
exports.HTTPRequest = HTTPRequest;
