// be-utils/lib/utils/LoggerFactory.js
/**
 *  Creator: Christian Hotz
 *  Company: hydra newmedia GmbH
 *  Date: 06.06.16
 *
 *  Copyright hydra newmedia GmbH
 */
"use strict";
/**
 *  Imports
 */
var winston = require('winston');
var winston_1 = require('winston');
var _ = require('lodash');
var LoggerFactory = (function () {
    function LoggerFactory(config) {
        this.config = config;
        this.loggers = {};
    }
    LoggerFactory.prototype.createWinstonLogger = function () {
        if (!_.has(this.loggers, 'winston')) {
            this.loggers['winston'] = new (winston_1.Logger)({
                transports: [
                    new (winston.transports.Console)({
                        json: false,
                        timestamp: true,
                        level: this.config.console.logLevel
                    }),
                    new winston.transports.File({
                        filename: this.config.baseDir + '/' + this.config.file.executionFile,
                        json: false,
                        level: this.config.file.logLevel
                    })
                ],
                exceptionHandlers: [
                    new (winston.transports.Console)({
                        json: true,
                        timestamp: true
                    }),
                    new winston.transports.File({
                        filename: this.config.baseDir + '/' + this.config.file.exceptionsFile,
                        json: false
                    })
                ],
                exitOnError: false
            });
        }
        return _.get(this.loggers, 'winston');
    };
    return LoggerFactory;
}());
exports.LoggerFactory = LoggerFactory;
